File: ios/A2dpPlaybackManager.swift
import AVFoundation

enum WearablesA2dpPlaybackState {
    nonisolated(unsafe) static var isActive = false
}

/// High-quality media playback to glasses over Bluetooth A2DP (output only).
/// See https://wearables.developer.meta.com/docs/develop/dat/microphones-and-speakers/
@MainActor
public final class A2dpPlaybackManager {
    public static let shared = A2dpPlaybackManager()

    private init() {}

    nonisolated static var isActive: Bool {
        WearablesA2dpPlaybackState.isActive
    }

    public func configure() throws -> [String: Any] {
        let session = AVAudioSession.sharedInstance()
        try session.setCategory(
            .playback,
            mode: .default,
            options: [.allowBluetoothA2DP]
        )
        return statusPayload()
    }

    @discardableResult
    public func activate() throws -> [String: Any] {
        if WearablesAudioSessionState.isActive {
            try AudioSessionManager.shared.deactivate()
        }

        let session = AVAudioSession.sharedInstance()
        try session.setCategory(
            .playback,
            mode: .default,
            options: [.allowBluetoothA2DP]
        )
        try session.setActive(true, options: .notifyOthersOnDeactivation)
        WearablesA2dpPlaybackState.isActive = true
        return statusPayload()
    }

    public func deactivate() throws {
        let session = AVAudioSession.sharedInstance()
        try session.setActive(false, options: .notifyOthersOnDeactivation)
        WearablesA2dpPlaybackState.isActive = false
    }

    private func statusPayload() -> [String: Any] {
        let outputs = AVAudioSession.sharedInstance().currentRoute.outputs
        let routedToA2dp = outputs.contains { $0.portType == .bluetoothA2DP }
        let a2dpAvailable = outputs.contains {
            $0.portType == .bluetoothA2DP || $0.portType == .bluetoothLE
        }
        let active = WearablesA2dpPlaybackState.isActive
        return [
            "active": active,
            "routedToBluetooth": active && routedToA2dp,
            "a2dpDeviceAvailable": a2dpAvailable
        ]
    }
}

File: android/src/main/java/expo/modules/emwdat/A2dpPlaybackManager.kt
package expo.modules.emwdat

import android.content.Context
import android.media.AudioDeviceInfo
import android.media.AudioManager

/**
 * Prepares high-quality media playback to glasses over Bluetooth A2DP (output only).
 * Use with expo-av, MediaPlayer, or AudioTrack after activate().
 * See https://wearables.developer.meta.com/docs/develop/dat/microphones-and-speakers/
 */
object A2dpPlaybackManager {
    private var isActive: Boolean = false

    fun isActive(): Boolean = isActive

    fun configure(context: Context): Map<String, Any> {
        val a2dpAvailable = findA2dpOutputDevice(context) != null
        return mapOf(
            "active" to isActive,
            "a2dpDeviceAvailable" to a2dpAvailable,
            "routedToBluetooth" to isActive && a2dpAvailable
        )
    }

    fun activate(context: Context): Map<String, Any> {
        if (AudioSessionManager.isActive()) {
            AudioSessionManager.deactivate(context)
        }

        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

        // Media playback uses MODE_NORMAL; release SCO if HFP was active.
        if (audioManager.isBluetoothScoOn) {
            audioManager.stopBluetoothSco()
            audioManager.isBluetoothScoOn = false
        }
        audioManager.mode = AudioManager.MODE_NORMAL

        val a2dpDevice = findA2dpOutputDevice(context)
        val routed = a2dpDevice != null
        isActive = routed

        return mapOf(
            "active" to isActive,
            "a2dpDeviceAvailable" to routed,
            "routedToBluetooth" to routed
        )
    }

    fun deactivate(context: Context) {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        audioManager.mode = AudioManager.MODE_NORMAL
        isActive = false
    }

    private fun findA2dpOutputDevice(context: Context): AudioDeviceInfo? {
        val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        return audioManager
            .getDevices(AudioManager.GET_DEVICES_OUTPUTS)
            .firstOrNull { it.type == AudioDeviceInfo.TYPE_BLUETOOTH_A2DP }
    }
}
